src/lib.rs

```diff
@@ -0,0 +1,83 @@
+use laz::las::file::{read_vlrs_and_get_laszip_vlr, QuickHeader};
+use laz::las::laszip::LasZipDecompressor;
+use wasm_bindgen::prelude::*;
+use std::io::Seek;
+extern crate console_error_panic_hook;
+
+#[wasm_bindgen]
+pub struct WasmQuickHeader {
+    pub major: u8,
+    pub minor: u8,
+    pub offset_to_points: u32,
+    pub num_vlrs: u32,
+    pub point_format_id: u8,
+    pub point_size: u16,
+    pub num_points: u64,
+    pub header_size: u16,
+}
+
+impl From<QuickHeader> for WasmQuickHeader {
+    fn from(other: QuickHeader) -> WasmQuickHeader {
+        WasmQuickHeader {
+            major: other.major,
+            minor: other.minor,
+            offset_to_points: other.offset_to_points,
+            num_vlrs: other.num_vlrs,
+            point_format_id: other.point_format_id,
+            point_size: other.point_size,
+            num_points: other.num_points,
+            header_size: other.header_size,
+        }
+    }
+}
+
+#[wasm_bindgen]
+pub fn get_header(buf: js_sys::Uint8Array)  -> std::result::Result<WasmQuickHeader, JsValue> {
+    // initialize debugging
+    console_error_panic_hook::set_once();
+
+    // copy header bytes into wasm memory
+    let mut body = vec![0; buf.length() as usize];
+    buf.copy_to(&mut body[..]);
+    // cursor to wrap the bytes
+    let mut cursor = std::io::Cursor::new(body);
+    let hdr = QuickHeader::read_from(&mut cursor).unwrap();
+    Ok(WasmQuickHeader::from(hdr))
+}
+
+#[wasm_bindgen]
+pub struct WasmLasZipDecompressor {
+    decompressor: LasZipDecompressor<'static, std::io::Cursor<Vec<u8>>>,
+}
+
+impl WasmLasZipDecompressor {
+    pub fn new(source: Vec<u8>) -> Self {        
```

Comment:
It think this could be `pub fn new(souce: &[u8]) -> Self { ... }` to avoid havinf to copy

Reply:
I still had to use `to_vec`, because the [docs](https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.Uint8Array.html#impl-15) say viewing the data directly is unsafe

Reply:
I still had to use `to_vec`, because the [docs](https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.Uint8Array.html#impl-15) say viewing the data directly is unsafe

Reply:
We are viewing data the other way around. We want to view JS memory as slice, not expose a Rust slice as JsUint8Array.

Just like we have a 
```Rust
pub fn decompress_many(&mut self, out: &mut [u8]) -> Result<(), JsValue>
```

and the user can give a `Uint8Array` as the `out` parameter, our `new` function could accept a `&[u8]` for which the user pass a `Uint8Array` as the source.

That would means we would need to change the WasmLasZipDecompressor, and make a few things more complex to make sure its correct and not unsafe so we can leave that for later.

Reply:
I think that's what I did originally. I don't remember why I had to change it. I'll look into it again, that would be the best solution.

Reply:
I remember the issue.

A lifetime specification is needed here:
> decompressor: LasZipDecompressor<'static, std::io::Cursor<&'static [u8]>>,

And as a result, I change the function signature as follows:

```
    pub fn new(buf: &'static [u8]) -> Result<WasmLasZipDecompressor, JsValue> {      
        let mut cursor = std::io::Cursor::new(buf);
```

However, this won't work because wasm-bindgen doesn't support lifetimes in signature. Similarly, we can't add a lifetime specification to WasmLasZipDecompressor. That's why I passed in the `Uint8Array` so that we didn't have to assume lifetime of the object. Let me know if you have any suggestions.

Reply:
Probably the safest solution is to create our own struct that wraps Uint8Array and implements the std::io::{Read, Seek} traits like std::io::Cursor;

Reply:
Yeah, lets leave it alone for now and we can merge. I'm curious how much overhead `to_vec` actually has, it seems like it doesn't actually copy the memory, just returns a view. But I'm not sure.